jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponse.java

```diff
@@ -249,136 +271,162 @@ protected void onCompleteFailure(Throwable x)
         @Override
         protected Action process() throws Exception
         {
-            // If we have no deflater
-            if (_deflaterEntry == null)
+            if (LOG.isDebugEnabled())
+                LOG.debug("GzipBufferCB.process(): _last={}, _buffer={}, _content={}", _last, BufferUtil.toDetailString(_buffer), BufferUtil.toDetailString(_content));
+
+            GZState gzstate = _state.get();
+
+            // Are we finished?
+            if (gzstate == GZState.FINISHED)
             {
                 // then the trailer has been generated and written below.
                 // We have finished compressing the entire content, so
                 // cleanup and succeed.
-                if (_buffer != null)
-                {
-                    getRequest().getComponents().getByteBufferPool().release(_buffer);
-                    _buffer = null;
-                }
-                if (_copy != null)
-                {
-                    getRequest().getComponents().getByteBufferPool().release(_copy);
-                    _copy = null;
-                }
+                cleanup();
                 return Action.SUCCEEDED;
             }
 
             // If we have no buffer
             if (_buffer == null)
             {
-                // allocate a buffer and add the gzip header.
                 _buffer = getRequest().getComponents().getByteBufferPool().acquire(_bufferSize, false);
-                BufferUtil.fill(_buffer, GZIP_HEADER, 0, GZIP_HEADER.length);
+                // Per RFC-1952, GZIP is LITTLE_ENDIAN
+                _buffer.order(ByteOrder.LITTLE_ENDIAN);
```

Comment:
This is true, but do we ever do non-byte operations on the buffer?

The endian state of a buffer from the pool perhaps is something we need to reset before putting back into the pool... except that we don't do non-byte operations.

Reply:
The buffer pool resets the order to BIG_ENDIAN on release already.

Reply:
Each of the ByteBufferPool implementations call `BufferUtil.reset(ByteBuffer)` on release.

Which does ...

https://github.com/eclipse/jetty.project/blob/d0cf73f1c7c7a37ad006abb0a6c52d3b37d6722f/jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java#L179-L187